---
title: "SCchicbackanalysis"
author: "Laurens-Willem Janssen"
date: "2023-07-13"
output:
  html_document: default
  word_document: default
  pdf_document: default
---
This is an R script made to perform a single-cell analysis on chicken skin back data.
Made by Laurens-Willem Janssen (s1114416) from Hogeschool Leiden.
For Leiden University, guided by Michael Richardson and Luthfi Nurhidayat.

Loading in the necessary packages.
```{r Load packages, message=FALSE, warning=FALSE}
library(Seurat, scater)
library(scran)
library(Matrix, magrittr)
lapply(c("dplyr", "HGNChelper", "openxlsx", "tidyverse", "data.tree", "igraph", "ggraph"), library, character.only = T)
```
Reading the Cell Ranger data.
```{r Read CR back data matrix, warning=FALSE}
back <- Read10X(data.dir = "filtered_feature_bc_matrixB")
back <- CreateSeuratObject(counts = back)
```
Adding the marker for the mitochondrial genes,
as it is not present in the raw Cell Ranger output.
```{r Add mitomarker}
mtlist = c("^ND1", "^ND2", "^COX1", "^COX2", "^ATP8", "^ATP6", "^COX3", "^ND3",
           "^ND4L", "^ND4", "^ND5", "^CYTB", "^ND6")
mtgene <- grep(paste(mtlist, collapse = "|"),
               back@assays[["RNA"]]@counts@Dimnames[[1]], value=TRUE)
mtgene <- paste(mtgene, collapse = "|")
```
Create and add the percentage of mitochondrial and ribosomal genes
```{r Percent mito and ribo}
percent.mito <- PercentageFeatureSet(back, pattern = mtgene)
back <- AddMetaData(back, percent.mito, col.name = "percent.mito")
percent.ribo <- PercentageFeatureSet(back, pattern = "^RP[SL]")
back <- AddMetaData(back, percent.ribo, col.name = "percent.ribo")
```
Deleting the percent.mito/ribo objects to save RAM & creating the scatter plots
```{r Scatterplot}
remove(percent.mito)
remove(percent.ribo)
remove(mtgene)
remove(mtlist)
table(Idents(back))
FeatureScatter(back, feature1 = "nCount_RNA", feature2 = "percent.mito") + NoLegend()
FeatureScatter(back, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
```
Cutting off over or under expressed genes and cells with the cutoff values based on the scatter plots made earlier.
```{r Apply cutoff}
back <- subset(back, subset = nFeature_RNA < 5000 & nFeature_RNA > 200 & nCount_RNA < 50000)
table(Idents(back))
```
Normalization and finding variable features.
```{r Normalization and variable}
back <- NormalizeData(back, normalization.method = "LogNormalize", scale.factor = 10000)
back <- FindVariableFeatures(back, selection.method = "vst", nfeatures = 2000)
```

Scale the data and perform the PCA.
```{r Data scaling & PCA}
allgenes <- rownames(back)
back <- ScaleData(back, features = allgenes)
remove(allgenes)
back <- RunPCA(back, features = VariableFeatures(object = back))
```
Performing the Jack Straw test on the PCA result.
```{r Jack Straw}
back <- JackStraw(back, num.replicate = 100)
back <- ScoreJackStraw(back, dims = 1:20)
```
Generate the elbow plot to select where the PC cutoff will be.
```{r Generate elbow plot}
ElbowPlot(back)
```
The elbow plot above tells that a cutoff around 10 PC would be the most suitable.
After some trying a cutoff of 11 ended up producing the best result.
A low resolution of 0.3 was chosen also, after some trial and error.
```{r Neighbors clusters & UMAP}
back <- FindNeighbors(back, dims = 1:11)
back <- FindClusters(back, resolution = 0.3)
back <- RunUMAP(back, dims = 1:11)
```
Creating the list of colors that will be used for visualizing the fourteen
generated clusters, and generating the UMAP showing the clusters.
```{r Set colors and create the UMAP clusterplot}
ccolss= c("#5f75ae","#92bbb8","#64a841","#e5486e","#de8e06","#eccf5a","#b5aa0f","#e4b680","#7ba39d","#b15928","#ffff99", "#6a3d9a","#cab2d6","#ff7f00","#fdbf6f","#e31a1c","#fb9a99","#33a02c","#b2df8a","#1f78b4","#a6cee3")
DimPlot(back, reduction = "umap", label = TRUE, cols = ccolss) + NoLegend()
```
Calculates what genes are the best markers for each cluster.
```{r Create markers}
markers <- FindAllMarkers(back, only.pos = TRUE, min.pct = 0, logfc.threshold = 0.25)
```
Making the heat map that shows the expression levels of the discovered
marker genes for each cluster.
```{r Create heat map}
markers %>%
  group_by(cluster) %>%
  top_n(n = 2, wt = avg_log2FC) -> topgene
DoHeatmap(back, features = topgene$gene, group.colors = ccolss, size = 4) + NoLegend()
```
Load the training and reference data needed for the cell type annotation with ScType.
```{r Load ScType data}
source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/gene_sets_prepare.R"); source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/sctype_score_.R")
gs_list = gene_sets_prepare("ScTypeDB_full.xlsx", "Chic")
db_ = "https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/ScTypeDB_full.xlsx";
```
The next commands are part of the ScType 'pipeline' for annotating the most characteristic cell type in each cluster.
```{r Generate score data}
es = sctype_score(scRNAseqData = back[["RNA"]]@scale.data, scaled = TRUE,
                  gs = gs_list$gs_positive, gs2 = gs_list$gs_negative)
```
```{r}
cL_resutls = do.call("rbind", lapply(unique(back@meta.data$seurat_clusters), function(cl){
  es = sort(rowSums(es[ ,rownames(back@meta.data[back@meta.data$seurat_clusters==cl, ])]), decreasing = !0)
  head(data.frame(cluster = cl, type = names(es), scores = es, ncells = sum(back@meta.data$seurat_clusters==cl)), 10)}))
sctype_scores = cL_resutls %>% group_by(cluster) %>% top_n(n = 1, wt = scores)
```
```{r}
sctype_scores$type[as.numeric(as.character(sctype_scores$scores)) < sctype_scores$ncells/4] = "Unknown"
print(sctype_scores[,1:3])
```
```{r Add to back data object}
back@meta.data$customclassif = ""
for(j in unique(sctype_scores$cluster)){
  cl_type = sctype_scores[sctype_scores$cluster==j,]; 
  back@meta.data$customclassif[back@meta.data$seurat_clusters == j] = as.character(cl_type$type[1])
}
```
Generate the UMAP showing the cell type for each data point.
```{r Create the UMAP cell type label plot}
DimPlot(back, reduction = "umap", label = TRUE, repel = TRUE, group.by = 'customclassif', cols = ccolss) + NoLegend()
```
The next commands are part of the process to create the bubble map, showing
the cell types discovered in each cluster.
```{r Prepare for bubble plot}
cL_resutls=cL_resutls[order(cL_resutls$cluster),]; edges = cL_resutls; edges$type = paste0(edges$type,"_",edges$cluster); edges$cluster = paste0("cluster ", edges$cluster); edges = edges[,c("cluster", "type")]; colnames(edges) = c("from", "to"); rownames(edges) <- NULL
```
```{r}
nodes_lvl1 = sctype_scores[,c("cluster", "ncells")]; nodes_lvl1$cluster = paste0("cluster ", nodes_lvl1$cluster); nodes_lvl1$Colour = "#f1f1ef"; nodes_lvl1$ord = 1; nodes_lvl1$realname = nodes_lvl1$cluster; nodes_lvl1 = as.data.frame(nodes_lvl1); nodes_lvl2 = c()
```
```{r}
for (i in 1:length(unique(cL_resutls$cluster))){
  dt_tmp = cL_resutls[cL_resutls$cluster == unique(cL_resutls$cluster)[i], ]; nodes_lvl2 = rbind(nodes_lvl2, data.frame(cluster = paste0(dt_tmp$type,"_",dt_tmp$cluster), ncells = dt_tmp$scores, Colour = ccolss[i], ord = 2, realname = dt_tmp$type))
}
```
```{r}
nodes <- rbind(nodes_lvl1, nodes_lvl2); nodes$ncells[nodes$ncells<1] = 1;
files_db = openxlsx::read.xlsx(db_)[,c("cellName", "shortName")]; files_db = unique(files_db); nodes = merge(nodes, files_db, all.x = T, all.y = F, by.x = "realname", by.y = "cellName", sort = F)
```
```{r}
nodes$shortName[is.na(nodes$shortName)] = nodes$realname[is.na(nodes$shortName)]; nodes = nodes[,c("cluster", "ncells", "Colour", "ord", "shortName", "realname")]
```
```{r}
mygraph <- graph_from_data_frame(edges, vertices=nodes)
```
generate the bubble plot.
```{r}
ggraph(mygraph, layout = 'circlepack', weight=I(ncells)) + 
  geom_node_circle(aes(filter=ord==1,fill=I("#F5F5F5"), colour=I("#D3D3D3")), alpha=0.9) + geom_node_circle(aes(filter=ord==2,fill=I(Colour), colour=I("#D3D3D3")), alpha=0.9) +
  theme_void() + geom_node_text(aes(filter=ord==2, label=shortName, colour=I("#ffffff"), fill="white", repel = !1, parse = T, size = I(log(ncells,25)*1.5)))+ geom_node_label(aes(filter=ord==1,  label=shortName, colour=I("#000000"), size = I(3), fill="white", parse = T), repel = !0, segment.linetype="dotted")
```
